File: apps/pos_app_legacy/app/src/main/kotlin/com/tienda/pos_legacy/model/Sale.kt
package com.tienda.pos_legacy.model

import org.json.JSONObject

data class SaleItemResult(
    val productId: String,
    val nombreProducto: String,
    val cantidad: Double,
    val precioUnitario: Double,
    val subtotal: Double,
) {
    companion object {
        fun fromJson(json: JSONObject): SaleItemResult {
            val product = json.optJSONObject("product")
            return SaleItemResult(
                productId = json.getString("productId"),
                nombreProducto = product?.optString("nombre") ?: "",
                cantidad = json.getDouble("cantidad"),
                precioUnitario = json.getDouble("precioUnitario"),
                subtotal = json.getDouble("subtotal"),
            )
        }
    }
}

data class SaleResult(
    val id: String,
    val numeroTicket: Int,
    val total: Double,
    val montoRecibido: Double?,
    val cambio: Double?,
    val fecha: String,
    val items: List<SaleItemResult>,
) {
    companion object {
        fun fromJson(json: JSONObject): SaleResult {
            val itemsJson = json.getJSONArray("items")
            val items = (0 until itemsJson.length()).map { SaleItemResult.fromJson(itemsJson.getJSONObject(it)) }
            return SaleResult(
                id = json.getString("id"),
                numeroTicket = json.getInt("numeroTicket"),
                total = json.getDouble("total"),
                montoRecibido = if (json.isNull("montoRecibido")) null else json.getDouble("montoRecibido"),
                cambio = if (json.isNull("cambio")) null else json.getDouble("cambio"),
                fecha = json.getString("fecha"),
                items = items,
            )
        }
    }
}

File: apps/pos_app/lib/models/sale.dart
class SaleItemResult {
  final String productId;
  final String nombreProducto;
  final double cantidad;
  final double precioUnitario;
  final double subtotal;

  SaleItemResult({
    required this.productId,
    required this.nombreProducto,
    required this.cantidad,
    required this.precioUnitario,
    required this.subtotal,
  });

  factory SaleItemResult.fromJson(Map<String, dynamic> json) {
    final product = json['product'] as Map<String, dynamic>?;
    return SaleItemResult(
      productId: json['productId'] as String,
      nombreProducto: product?['nombre'] as String? ?? '',
      cantidad: (json['cantidad'] as num).toDouble(),
      precioUnitario: (json['precioUnitario'] as num).toDouble(),
      subtotal: (json['subtotal'] as num).toDouble(),
    );
  }
}

class SaleResult {
  final String id;
  final int numeroTicket;
  final double subtotal;
  final double descuento;
  final double total;
  final String medioPago;
  final double? montoRecibido;
  final double? cambio;
  final DateTime fecha;
  final List<SaleItemResult> items;

  SaleResult({
    required this.id,
    required this.numeroTicket,
    required this.subtotal,
    required this.descuento,
    required this.total,
    required this.medioPago,
    required this.montoRecibido,
    required this.cambio,
    required this.fecha,
    required this.items,
  });

  factory SaleResult.fromJson(Map<String, dynamic> json) {
    return SaleResult(
      id: json['id'] as String,
      numeroTicket: json['numeroTicket'] as int,
      subtotal: (json['subtotal'] as num).toDouble(),
      descuento: (json['descuento'] as num).toDouble(),
      total: (json['total'] as num).toDouble(),
      medioPago: json['medioPago'] as String,
      montoRecibido: (json['montoRecibido'] as num?)?.toDouble(),
      cambio: (json['cambio'] as num?)?.toDouble(),
      fecha: DateTime.parse(json['fecha'] as String),
      items: (json['items'] as List<dynamic>)
          .map((e) => SaleItemResult.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }
}
